rust/src/composepost.rs

```diff
@@ -0,0 +1,98 @@
+/*
+ * Copyright (C) 2018 Red Hat, Inc.
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation; either version 2 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA
+ */
+
+use openat;
+use std::{io, fs};
+use std::io::{Write, BufRead};
+use std::path::Path;
+use failure::Fallible;
+
+// IMO should propose this at least in a "utils" bit of the openat crate;
+// Like 95% of the time I'm looking at errno (with files) it's for ENOENT,
+// and Rust has an elegant way to map that with Option<>.  Every other
+// error I usually just want to propagate back up.
+fn openat_optional<P, F>(d: &openat::Dir, p: P, f: F) -> io::Result<Option<fs::File>>
+    where P: openat::AsPath, F: FnOnce(&openat::Dir, P) -> io::Result<fs::File>
```

Comment:
Why pass in a function here? This is only really useful for `open_file`, right? E.g. `{write,append,create}_file` all create the file it it doesn't exist.

Reply:
I think openat is missing the equivalent of `O_APPEND` without `O_CREAT` for example.  But...that's really obscure I guess, and even if it wasn't we'd probably want a wrapper anyways.